src/vfs.c

```diff
@@ -1108,7 +990,29 @@ static int vfsWalFileRead(sqlite3_file* file, void* buf, int amount, sqlite3_int
 		return SQLITE_IOERR_SHORT_READ;
 	}
 
-	frame = vfsWalFrameLookup(f->wal, index);
+	/* If ever really needed, there exists lock-free algorithms. I think
+	 * that here contention will always be very small the only thing that
+	 * can be contended is the access to the wal page pointer and not
+	 * on the content: it is guaranteed by the sqlite3 WAL protocol that
+	 * the WAL will not be restarted or checkpointed if a READ lock is
+	 * held, which is always true when reading the WAL. */
+	mtx_lock(&f->database->lock);
```

Comment:
Could you explain in a comment around the database lock itself, that this lock appeared because of an added source of parallelism and as some of vfs functions are called from new threads, uv thread needs a certain serialization with others.